buckets/local/watch.go

```diff
@@ -5,92 +5,175 @@ import (
-	args := &watchOptions{
-		interval: defaultInterval,
-	}
-	for _, opt := range opts {
-		opt(args)
-	}
-	errs := make(chan error)
+// WatchState is used to inform Watch callers about the connection state.
+type WatchState struct {
+	// State of the watch connection (online/offline).
+	State ConnectionState
+	// Error returned by the watch operation.
+	Err error
+	// Fatal indicates whether or not the associated error is fatal.
+	// (Connectivity related errors are not fatal.)
+	Fatal bool
+}
 
-	// Manually sync once on startup
-	if err := b.watchPush(ctx, args.events); err != nil {
-		return err
-	}
+// ConnectionState indicates an online/offline state.
+type ConnectionState int
 
-	// Start listening for local changes
-	w := watcher.New()
-	defer w.Close()
-	w.SetMaxEvents(1)
-	bp, err := b.Path()
-	if err != nil {
-		return err
-	}
-	if err := w.AddRecursive(bp); err != nil {
-		return err
+const (
+	// Offline indicates the remote is currently not reachable.
+	Offline ConnectionState = iota
+	// Online indicates a connection with the remote has been established.
+	Online
+)
+
+func (cs ConnectionState) String() string {
+	switch cs {
+	case Online:
+		return "online"
+	case Offline:
+		return "offline"
+	default:
+		return "unknown state"
 	}
+}
+
+// Watch watches for and auto-pushes local bucket changes at an interval,
+// and listens for and auto-pulls remote changes as they arrive.
+// This will survive connectivity interruptions.
+// Returns a channel of watch connectivity states. Cancel context to stop watching.
+func (b *Bucket) Watch(ctx context.Context, opts ...WatchOption) <-chan WatchState {
+	bc := backoff.NewConstantBackOff(reconnectInterval)
+	outerState := make(chan WatchState)
 	go func() {
-		if err := w.Start(args.interval); err != nil {
-			errs <- err
-		}
-	}()
-	go func() {
-		for {
-			select {
-			case <-w.Event:
-				if err := b.watchPush(ctx, args.events); err != nil {
-					errs <- err
+		defer close(outerState)
+		err := backoff.Retry(func() error {
+			state, err := b.WatchWhileConnected(ctx, opts...)
+			if err != nil {
+				outerState <- WatchState{Err: err, Fatal: true}
+				return nil // Stop retrying
+			}
+			for s := range state {
+				outerState <- s
+				if s.Err != nil {
+					if s.Fatal {
+						return nil // Stop retrying
+					} else {
+						return s.Err // Connection error, keep trying
+					}
 				}
-			case err := <-w.Error:
-				errs <- err
-			case <-w.Closed:
-				return
 			}
+			return nil
+		}, bc)
+		if err != nil {
+			outerState <- WatchState{Err: err, Fatal: true}
 		}
 	}()
+	return outerState
+}
 
-	// Start listening for remote changes
+// WatchWhileConnected watches for and auto-pushes local bucket changes at an interval,
+// and listens for and auto-pulls remote changes as they arrive.
+// This will watch until context is canceled or an error occurs.
+func (b *Bucket) WatchWhileConnected(ctx context.Context, opts ...WatchOption) (<-chan WatchState, error) {
```

Comment:
Nit: Maybe consolidate both `Watch*` APIs in a single one with an option? (e.g: `WithRetry(bool)`).

Reply:
Nice, like it